lib/internal/repl.js

```diff
@@ -192,31 +188,49 @@ function setupHistory(repl, historyPath, oldHistoryPath, ready) {
     timer = setTimeout(flushHistory, kDebounceHistoryMS);
   }
 
+  const historyLockFile = `${historyPath}.LCK`;
   function flushHistory() {
     timer = null;
     if (writing) {
       pending = true;
       return;
     }
-    writing = true;
-    const historyData = repl.history.join(os.EOL);
-    fs.write(repl._historyHandle, historyData, 0, 'utf8', onwritten);
-  }
-
-  function onwritten(err, data) {
-    writing = false;
-    if (pending) {
-      pending = false;
-      online();
-    } else {
-      repl._flushing = Boolean(timer);
-      if (!repl._flushing) {
-        repl.emit('flushHistory');
+    // Ref: https://github.com/nodejs/node/issues/1634
+    // opening a file for exclusive write should be atomic
+    // on non-networked POSIX systems.
+    fs.open(historyLockFile, 'wx', (err, fd) => {
+      if (err && err.code === 'EEXIST') {
+        pending = true;
+        repl.resume();
+        return;
       }
-    }
+      writing = true;
+      const lastLine = repl.history[0] + '\n';
+      fs.write(repl._historyHandle, lastLine, NaN, 'utf8', onwritten(fd));
+    });
   }
-}
 
+  function onwritten(fd) {
+    return (err, data) => {
+      writing = false;
+      if (pending) {
+        pending = false;
+        online();
+      } else {
+        repl._flushing = Boolean(timer);
+        if (!repl._flushing) {
+          repl.emit('flushHistory');
+        }
+      }
+      unlockHistoryFile(fd);
```

Comment:
> Hmm - I can only reproduce this when there is already a stray `.node_repl_history.LCK` file at start up, and the history file is never written to.

Well, yeah, this is how I can reproduce it:
```bash
$ ./node
> .exit
$ ./node
> .exit
[hangs]
```

The first one here leaves the stray lock file lying around.

This is just an idea, but how about this approach when saving the history:
1. Read the current history file into memory
2. Remove it
3. Create the history file again, with flags set to `wx`. If that fails, go back to 1.
4. Write the desired number of lines to the opened history file

That seems to be free of race conditions and does not require a lock file that could keep lying around if the process crashes (for whatever reason, it doesn’t have to be REPL-related)… what do you think?